copy: avoid using all system memory with authz plugins

runcom opened the pull request:
Authz plugins proxy the whole response to be later inspected by the
plugin in a byte slice (using append). The code uses io.Copy() directly
causing the whole response to be buffered in the byte slice in the
`responseModifier` of the authz sdk.
I noticed `docker save` doesn't have the same issue even streaming huge
tarballs by using ioutils.WriteFlusher.
This patch changes the copy code to:

- Use ioutils.WriteFlusher
- Use pools.Copy

We cannot reproduce the OOM anymore with this.

@cpuguy83 PTAL

Fix https://github.com/moby/moby/issues/36576

Signed-off-by: Antonio Murdaca <runcom@redhat.com>

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

cpuguy83 commented:
Wow, thanks for looking at this.

So is the issue that the write buffer filling up?
`io.Copy` should only have a 32K buffer as is.

Can you think of some way to unit test this?

tonistiigi commented:
SGTM

runcom commented:
> Can you think of some way to unit test this?

I've spent yesterday to think of something but couldn't come up with anything lol, do you have ideas?

runcom commented:
are test failures related? looks like they're not

ripcurld commented:
It looks like unrelated. `janky` was aborted (probably timeout) and others failed in different tests.
Let's try again. :smiley:

cpuguy83 commented:
@runcom What about something like this: https://play.golang.org/p/G1BWKP9Xvut

ripcurld commented:
@runcom it seems like this change makes `TestExternalVolumeDriverUnmountOnCp` fail.

```
15:01:07 FAIL: docker_cli_external_volume_driver_unix_test.go:621: DockerExternalVolumeSuite.TestExternalVolumeDriverUnmountOnCp
15:01:07 
15:01:07 [d868eebd8df13] waiting for daemon to start
15:01:07 [d868eebd8df13] daemon started
15:01:07 
15:01:07 docker_cli_external_volume_driver_unix_test.go:630:
15:01:07     c.Assert(s.ec.unmounts, checker.Equals, 1, check.Commentf(out))
15:01:07 ... obtained int = 0
15:01:07 ... expected int = 1
15:01:07 ... 
15:01:07 
15:01:07 [d868eebd8df13] exiting daemon
```

kolyshkin commented:
CI failure on **experimental** ([logs](https://jenkins.dockerproject.org/job/Docker-PRs-experimental/39822/console)) is unrelated; see https://github.com/moby/moby/issues/36547

CI failure on **powerpc** ([logs](https://jenkins.dockerproject.org/job/Docker-PRs-powerpc/8998/console)) is interesting:

10:17:16 === RUN   TestMountDaemonRoot/private/mount_root
10:17:16 === RUN   TestMountDaemonRoot/private/mount_subpath
10:17:19 panic: Error response from daemon: removal of container 9588fac23c7077479cff76aa7e1785687c2829aacf5067b861ccc7d609bfeaf0 is already in progress [recovered]
10:17:19 	panic: Error response from daemon: removal of container 9588fac23c7077479cff76aa7e1785687c2829aacf5067b861ccc7d609bfeaf0 is already in progress
10:17:19 
10:17:19 goroutine 655 [running]:
10:17:19 testing.tRunner.func1(0xc420509c20)
10:17:19 	/usr/local/go/src/testing/testing.go:711 +0x290
10:17:19 panic(0x103e8b00, 0xc4204f8eb0)
10:17:19 	/usr/local/go/src/runtime/panic.go:491 +0x244
10:17:19 github.com/docker/docker/integration/container.TestMountDaemonRoot.func1.1.1(0x106f4380, 0xc42011ba00, 0x106e9580, 0xc42001a078, 0xc4200147c0, 0x40, 0x0, 0x0, 0x0)
10:17:19 	/go/src/github.com/docker/docker/integration/container/mounts_linux_test.go:261 +0xc4
10:17:19 github.com/docker/docker/integration/container.TestMountDaemonRoot.func1.1(0xc420509c20)
10:17:19 	/go/src/github.com/docker/docker/integration/container/mounts_linux_test.go:277 +0x558
10:17:19 testing.tRunner(0xc420509c20, 0xc4201a3f20)
10:17:19 	/usr/local/go/src/testing/testing.go:746 +0xc0
10:17:19 created by testing.(*T).Run
10:17:19 	/usr/local/go/src/testing/testing.go:789 +0x264

The other two failures (on **janky** and **z**) look like being genuine:

11:49:44 
11:49:44 ----------------------------------------------------------------------
11:49:44 FAIL: docker_cli_external_volume_driver_unix_test.go:621: DockerExternalVolumeSuite.TestExternalVolumeDriverUnmountOnCp
11:49:44 
11:49:44 [ddf1a70450757] waiting for daemon to start
11:49:44 [ddf1a70450757] daemon started
11:49:44 
11:49:44 docker_cli_external_volume_driver_unix_test.go:630:
11:49:44     c.Assert(s.ec.unmounts, checker.Equals, 1, check.Commentf(out))
11:49:44 ... obtained int = 0
11:49:44 ... expected int = 1
11:49:44 ... 
11:49:44 
11:49:44 [ddf1a70450757] exiting daemon

kolyshkin commented:
Oh the powerpc panic on test seems to be deliberate (see commit 589a0afa8c from PR #36055). @cpuguy83 care to take a look (logs are here: https://jenkins.dockerproject.org/job/Docker-PRs-powerpc/8998/console)?

cpuguy83 commented:
Not sure why I put a panic in there... should just fail the test or something. Also really weird error since we only call remove once.

cpuguy83 commented:
Should we set a max buffer size in the authz buffer?

runcom commented:
> Should we set a max buffer size in the authz buffer?

yeah, not sure what amount though, hint?

cpuguy83 commented:
How about 1MB?

cpuguy83 commented on api/server/router/container/copy.go:
```diff
@@ -103,7 +105,9 @@ func writeCompressedResponse(w http.ResponseWriter, r *http.Request, body io.Rea
 		cw = w
 	}
-	_, err := io.Copy(cw, body)
+	output := ioutils.NewWriteFlusher(cw)
```
This doesn't work. The compressed writer is not a flusher.

runcom replied:
:confused:

cpuguy83 replied:
I pushed an update that fixes this, adds a test, and a max buffer for authz responses.

cpuguy83 commented:
And the 1MB buffer limit highlighted that `docker export` was also affected (authz integration tests caught it).

Unfortunately triggering the error can be quite difficult without making the buffer smaller. The integration test I added for container copy to/from can't trigger it without making a huge file (10GB worked, 500MB did not, both are really too big to do in CI), but at least it makes sure copy works with authz, and then the separate unit test on the container copy stream writer to make sure there isn't any internal buffering.

cpuguy83 commented:
But now export container, export image, and container copy all seem to be setup correctly.

cpuguy83 commented:
Ouch, before this PR:

> PASS: docker_cli_external_volume_driver_unix_test.go:621: DockerExternalVolumeSuite.TestExternalVolumeDriverUnmountOnCp	1.434s

After this PR:

> PASS: docker_cli_external_volume_driver_unix_test.go:709: DockerExternalVolumeSuite.TestExternalVolumeDriverUnmountOnCp	31.727s

cpuguy83 committed:
Limit authz response buffer

When the authz response buffer limit is hit, perform a flush.
This prevents excessive buffer sizes, especially on large responses
(e.g. `/containers/<id>/archive` or `/containers/<id>/export`).

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 commented:
Ok, I've updated this (again, sorry...) with a more complete fix.
Now when the authz buffer limit is reached (currently set to a 64K limit) the buffer is flushed.

vdemeester reviewed (approved):
LGTM :tiger:

runcom commented:
:+1:

tonistiigi commented:
LGTM
